include/fiction/networks/virtual_pi_network.hpp

```diff
@@ -0,0 +1,523 @@
+    [[nodiscard]] bool is_virtual_ci(node const& n) const
+    {
+        return std::find(v_storage->virtual_inputs.cbegin(), v_storage->virtual_inputs.cend(), n) !=
+               v_storage->virtual_inputs.cend();
+    }
+
+    /**
+     * Check if a given node is a real CI in the virtual_pi_network.
+     *
+     * @param n The node to check.
+     * @return True if the node is a real CI, false otherwise.
+     */
+    [[nodiscard]] bool is_real_ci(node const& n) const
+    {
+        return (Ntk::is_ci(n) && !is_virtual_ci(n));
+    }
+
+    /**
+     * Get the number of virtual CIs in the virtual_pi_network.
+     *
+     * @return The number of virtual CIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_virtual_cis() const
+    {
+        return static_cast<uint32_t>(v_storage->virtual_inputs.size());
+    }
+
+    /**
+     * Get the number of real CIs in the virtual_pi_network.
+     *
+     * @return The number of real CIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_real_cis() const
+    {
+        return static_cast<uint32_t>(Ntk::num_cis() - num_virtual_cis());
+    }
+
+    /**
+     * Get the number of virtual PIs in the virtual_pi_network.
+     *
+     * @return The number of virtual PIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_virtual_pis() const
+    {
+        return static_cast<uint32_t>(v_storage->virtual_inputs.size());
+    }
+
+    /**
+     * Get the number of real PIs in the virtual_pi_network.
+     *
+     * @return The number of real PIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_real_pis() const
+    {
+        return static_cast<uint32_t>(Ntk::num_pis() - num_virtual_pis());
+    }
+
+    /**
+     * Get the real PI associated with a virtual PI node.
+     *
+     * @param v_pi The virtual pi node to retrieve the real pi for.
+     * @return The real pi associated with the virtual pi node.
+     */
+    [[nodiscard]] auto get_real_pi(const node& v_pi) const
+    {
+        const auto it = v_storage->map_virtual_to_real_pi.find(v_pi);
+
+        assert(it != v_storage->map_virtual_to_real_pi.end() && "Error: node is not a virtual pi");
+
+        return it->second;
+    }
+
+    /**
+     * Iterates over the real PIs of the circuit and applies a given function.
+     *
+     * @tparam Fn The type of the function to be applied.
+     * @param fn The function to be applied.
+     */
+    template <typename Fn>
+    void foreach_real_pi(Fn&& fn) const
+    {
+        static_cast<const Ntk*>(this)->foreach_pi(
+            [&](const auto& i)
+            {
+                if (!is_virtual_pi(i))
+                {
+                    std::forward<Fn>(fn)(i);
+                }
+            });
+    }
+
+    /**
+     * Iterates over the virtual PIs of the circuit and applies a given function.
+     *
+     * @tparam Fn The type of the function to be applied.
+     * @param fn The function to be applied.
+     */
+    template <typename Fn>
+    void foreach_virtual_pi(Fn&& fn) const
```

Comment:
I just copied this from mockturtle. There it also gives this clang-tidy issue. I don't know if this issue is to be taken seriousely. Maybe @marcelwa you have a comment on this.

Reply:
This is indeed an oversight on mockturtle's end. It can be easily fixed (and should be) like this:
```c++
template <typename Fn>
    void foreach_virtual_pi(Fn&& fn) const
    {
        mockturtle::detail::foreach_element(v_storage->virtual_inputs.cbegin(), v_storage->virtual_inputs.cend(), std::forward<Fn>(fn));
    }
```